File: Python/InstrumentService.py
# instrument_service.py
import json
import uuid
import websockets
import sys
from websocket import create_connection
from PopUpService import PopUpService
from Models.InstrumentInfo import InstrumentInfo

class InstrumentService:
    def get_instrument_details(self, uri):
        try:
            ws = create_connection(uri)
            print("Connected to WebSocket for instrument name.")

            request = {
                "type": "get",
                "group": "thirdPartyAPI",
                "data": {
                        "field": "information",
                        "value": True
                        },
                "messageID": str(uuid.uuid4())
                }
            json_request =  json.dumps(request)
            ws.send(json_request)


            response = ws.recv()
            
            json_response = json.loads(response)
            
            instrument_value = json_response.get('data', {}).get('value', {})
            
            if instrument_value==True:

                raise json.JSONDecodeError("Failed to decode JSON response", response, 0)
                
            instrument_info = InstrumentInfo(
            instrument_name=instrument_value.get('instrumentName', 'Unknown Instrument'),
            serial_number=instrument_value.get('serialNumber')
            )
            print(f'Instrument Name: {instrument_info.instrument_name}, Instrument Serial: {instrument_info.serial_number}')
            ws.close()
            return instrument_info

        except json.JSONDecodeError as json_ex:
            print(f"JSON error occurred: {json_ex}")
            PopUpService.show_popup("Warning", "Invalid Login to LMS Exchange. The ID/Password is incorrect or the user does not have sufficient privileges.")
            sys.exit()
        except Exception as ex:
            print(f"An unexpected error occurred: {ex}")
            PopUpService.show_popup("Warning", "An unexpected error occurred. Please try again. From Instrument")
            sys.exit()

File: C#/InstrumentService.cs
using System;
using System.Text.Json;
using System.Threading.Tasks;
using System.Net.WebSockets;

public class InstrumentService
{
    public async Task<InstrumentInfo> GetInstrumentDetailsAsync(string uri)
    {
        try
        {
            using (var webSocketClient = new WebSocketClient())
            {
                await webSocketClient.ConnectAsync(uri);
                Console.WriteLine("Connected to WebSocket for instrument name.");

                var request = new
                {
                    type = "get",
                    group = "thirdPartyAPI",
                    data = new
                    {
                        field = "information",
                        value = true
                    },
                    messageID = Guid.NewGuid().ToString()
                };

                string jsonRequest = JsonSerializer.Serialize(request);
                await webSocketClient.SendAsync(jsonRequest);

                string response = await webSocketClient.ReceiveAsync();
                Console.WriteLine("Raw JSON response: " + response);

                var jsonResponse = JsonSerializer.Deserialize<JsonResponse>(response);
                var instrumentValue = jsonResponse?.data?.value;
                
                var instrumentInfo = new InstrumentInfo             // Create an instance of InstrumentInfo
                {
                    InstrumentName = instrumentValue?.instrumentName ?? "Unknown Instrument",
                    SerialNumber = instrumentValue?.serialNumber
                };

            return instrumentInfo; // Now return the InstrumentInfo object
            }
        }
        catch (JsonException jsonEx)
        {
            Console.WriteLine($"JSON error occurred: {jsonEx.Message}");
            PopupService.ShowPopup("Invalid Login. The ID/Password is incorrect or the user does not have sufficient privileges", "Warning");
            Environment.Exit(1);
            return null;
        }
        catch (WebSocketException webSocketEx)
        {
            Console.WriteLine($"WebSocket error occurred: {webSocketEx.Message}");
            PopupService.ShowPopup("WebSocket connection failed.", "Warning");
            Environment.Exit(1);
            return null;
        }
        catch (Exception ex)
        {
            Console.WriteLine($"An unexpected error occurred: {ex.Message}");
            PopupService.ShowPopup("An unexpected error occurred. Please try again.", "Warning");
            Environment.Exit(1);
            return null;
        }
    }
}